Clean up backend subscriber code.

showell opened the pull request:
This sets the stage for us to use better data structures over the wire.

timabbott commented on static/js/stream_data.js:
```diff
@@ -216,8 +229,8 @@ exports.add_sub = function (sub) {
-        sub.subscribers = new LazySet([]);
+    if (!stream_subscribers.has(sub.stream_id)) {
+        exports.set_subscribers(sub, []);
     }
```
I guess the theory behind this is that it must be a "new" stream if we don't already have subscriber data for it?  

(I know this block later moves, and this function is only used in tests; I'm msotly wondering if we need more of a comment here).

showell replied:
This code is still in a bit of flux, so I will try to improve comments around `peer_data` once everything stabilizes.  For now we at least have this:

~~~
 11 // This maps a stream_id to a LazySet of user_ids who are subscribed.
 12 // We maintain the invariant that this has keys for all all stream_ids
 13 // that we track in the other data structures.  We initialize it during
 14 // clear_subscriptions.

timabbott commented:
This looks great!  I merged the frontend half of this as the series through 2382fa7a19b0e2a9a0d6e59a8f0070754dadc19f; looking at the rest now.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -2666,14 +2666,14 @@ def validate_user_access_to_subscribers_helper(
-        # streams in the first place.
-
+        else:
+            raise JsonableError(_("Subscriber data is not available for this stream"))
```
Do we need this line?  I think falling through, the `user_profile.can_access_public_streams` is equivalent to checking if the user is a guest, right?

showell replied:
I think explicit is better here, but I added a commit instead that just makes the comment more accurate about the current code structure.  (I haven't pushed yet, still dealing with other comments.)

timabbott commented on zerver/lib/actions.py:
```diff
@@ -5050,15 +5050,6 @@ def build_stream_dict_for_sub(
-    # Guest users lose access to subscribers when they are unsubscribed if the stream
-    # is not web-public.
-    if not sub["active"] and user.is_guest and not stream["is_web_public"]:
-        subscribers = None
```
We should probably document the (subtle) API change here in the API changelog.

timabbott replied:
I think there's a sensible argument that we should communicate "no data provided" for `sub.subscribers` differently from "there are no subscribers"; this changes causes us to send `[]` for "no data provided" instead of `None`.  I think it's equivalent conceptually -- Zulip's invariants currently don't allow non-deactivated streams without any subscribers; but it would somewhat box us in from changing that model in the future, because clients like the mobile apps would be treating `[]` as "no data" rather than "no subscribers".  Maybe not a corner case worth worrying too much about, especially since the mobile apps don't fetch this data set and may never do so.

@gnprice thoughts on that question?  It certainly simplifies this PR to switch from "no `subscribers` field" to `[]` for cases where due to permissions we're not providing subscriber data, but I think following the refactoring in this PR, we could easily enough get back the old behavior if that were preferred.  (See the commit message for discussion of how the webapp and ZT work with this currently).

timabbott replied:
I'm potentially happy for us to decide this by just trying to write the documentation explaining this API and decide whether on reading that we feel we've made things more or less simple to understand.

showell replied:
Just to be clear, the current situation on master is that we **randomly** either use `[]` or omit the field altogether, and this dates back to this commit from Tim in 2013: 5e420dc5c3b0749a1da629502ab140e5cb9c50ce

So the appropriate fix here is to make it consistent.  Omitting the field presents all kinds of practical problems for downstream code, and it puts us in a position where people might come to rely on the presence/absence of the field to determine whether a user can access subscribers, which I think would be brittle.  And as I mention in the commit message, the current approach taken by ZT and the webapp is to calculate access via other means.

So my plan here is to just bump the API version here and tweak the docs.

showell replied:
One of the places that benefits from having an array is `apply_event`, by the way.  My fix probably cleans up some obscure bugs where guest users would cause tracebacks in `apply_event`, since some of the fix-up code assumes `subscribers` is always there.

The context of this PR is also kind of important--the over-arching goal here is to take `subscribers` off of the "sub" objects in the **backend** (similar to what happens in the frontend now due to the commits that you merged yesterday).  So the whole mechanism is gonna change for the webapp.  The webapp will internally change how we call `fetch_initial_state_data` so that subscriber info is put into a separate data structure where user_ids map to list of stream_ids and thus provide better compression.

timabbott replied:
sgtm given it was random since 2013.

showell commented:
I believe I have addressed all the comments here, so I'm ready for the next round of review.

These are hopefully the last prep commits before the actual changes here, which I am still polishing up on my local branch.

timabbott commented:
Merged, thanks @showell!
